pytorch_lightning/core/step_result.py

```diff
@@ -0,0 +1,327 @@
+from typing import Optional, Dict, Union
+from torch import Tensor
+import torch
+from copy import copy
+
+
+class Result(Dict):
+
+    def __init__(
+            self,
+            minimize: Optional[Tensor] = None,
+            early_stop_on: Tensor = None,
+            checkpoint_on: Union[Tensor, bool] = None,
+            hiddens: Optional[Tensor] = None
+    ):
+
+        super().__init__()
+
+        if early_stop_on is not None:
+            self.early_stop_on = early_stop_on
+        if checkpoint_on is not None and checkpoint_on:
+            self.checkpoint_on = checkpoint_on
+        if hiddens is not None:
+            self.hiddens = hiddens
+        if minimize is not None:
+            err = 'Minimize can only be used in training_end, training_step_end, training_epoch_end'
+            self._assert_grad_tensor_metric('minimize', minimize, err)
+            self.minimize = minimize
+
+        if minimize is not None and checkpoint_on is None:
+            self.checkpoint_on = minimize.detach()
+
+        self['meta'] = {
+            '_internal': {
+                '_reduce_on_epoch': False
+            }
+        }
+
+    def __getattr__(self, key):
+        try:
+            if key == 'callback_metrics':
+                return self.get_callback_metrics()
+            elif key == 'batch_log_metrics':
+                return self.get_batch_log_metrics()
+            elif key == 'batch_pbar_metrics':
+                return self.get_batch_pbar_metrics()
+            elif key == 'epoch_log_metrics':
+                return self.get_epoch_log_metrics()
+            elif key == 'epoch_pbar_metrics':
+                return self.get_epoch_pbar_metrics()
+            else:
+                return self[key]
+        except KeyError:
+            return None
+
+    def __setattr__(self, key, val):
+        # ensure reserve keys are tensors and detached
```

Comment:
```suggestion
        # ensure reserve keys are tensors and detached
        if isinstance(val, torch.Tensor) and key != 'minimize':
            val = val.detach()
            
        if key in {'hiddens', 'checkpoint_on', 'early_stop_on'}:
            self._assert_tensor_metric(key, val)
```

combine detaching logic

Reply:
@justusschock could you pls use multi-line suggestions... :]